src/main/java/jenkins/branch/MultiBranchProjectDescriptor.java

```diff
@@ -56,6 +60,83 @@
 public abstract class MultiBranchProjectDescriptor extends AbstractFolderDescriptor {
 
     /**
+     * The base class of the projects that are produced by this factory.
+     *
+     * @since 2.0
+     */
+    @NonNull
+    private final Class<? extends Job> projectClass;
+
+    /**
+     * Explicit constructor to use when type inference fails.
+     *
+     * @param clazz        the {@link MultiBranchProject} that this descriptor is for.
+     * @param projectClass the {@link Job} type that the {@link MultiBranchProject} creates.
+     * @since 2.0
+     */
+    protected MultiBranchProjectDescriptor(Class<? extends MultiBranchProject<?, ?>> clazz,
+                                           Class<? extends Job> projectClass) {
+        super(clazz);
+        this.projectClass = projectClass;
+    }
+
+    /**
+     * Semi explicit constructor to use when the descriptor is not an inner class of the {@link MultiBranchProject}.
+     *
+     * @param clazz the {@link MultiBranchProject} that this descriptor is for.
+     * @since 2.0
+     */
+    protected MultiBranchProjectDescriptor(Class<? extends MultiBranchProject<?, ?>> clazz) {
+        super(clazz);
+        projectClass = inferProjectClass(clazz);
+    }
+
+    /**
+     * Fully inferring constructor to use when the descriptor is an inner class of the {@link MultiBranchProject}
+     * and type parameter inference works because it just should work.
+     *
+     * @since 2.0
+     */
+    protected MultiBranchProjectDescriptor() {
```

Comment:
Again please make sure you test against: `MatrixMultiBranchProject`, `MavenMultiBranchProject`, `FreeStyleMultiBranchProject`, `IvyMultiBranchProject`, `YamlMultiBranchProject`, `PipelineMultiBranchDefaultsProject`, `FreeStyleMultiBranchProject`, `WorkflowMultiBranchProject`, `LiterateMultibranchProject`